[MP Rewrite Part 3] Add PlanRequest and the planner node.

kylestach opened the pull request:
This also overhauls the MotionCommand structure.

kylestach committed:
Add PlanRequest and the planner node.

This also overhauls the MotionCommand structure.

kylestach committed:
Merge branch 'mp-rewrite/trajectory' into mp-rewrite/plan-request

oswinso commented on soccer/Processor.cpp:
```diff
@@ -252,6 +252,11 @@ void Processor::run() {
         // TODO(Kyle): This is dead code for now. Once everything is ported over
         // to modules we can delete the if (false), but for now we still have to
         // update things manually.
+#if 0
+        for (auto* node : _nodes) {
+            node->run();
+        }
+#endif
```
I think you should just remove this directly

oswinso commented on soccer/planning/planner/MotionCommand.hpp:
```diff
@@ -0,0 +1,70 @@
+ */
+struct EmptyCommand {};
+
+/**
+ * Move to a particular target with a particular velocity, avoiding obstacles.
+ */
+// note: as of now the heading and angular velocity are ignored
+// TODO(Kyle) use heading and angular velocity or change this to a
+// LinearMotionInstant
```
Issue num

kylestach replied:
#1505

oswinso commented on soccer/planning/planner/MotionCommand.hpp:
```diff
@@ -0,0 +1,70 @@
+#include "planning/Trajectory.hpp"
+
+namespace Planning {
+// todo(Ethan) discuss whether we want to have settle with bounce
```
Issue num

kylestach replied:
This is resolved, I'll delete it.

oswinso reviewed (changes requested):
More issue nums

kylestach committed:
Remove dead code from Processor, replace with TODO

kylestach committed:
Make path target goal a linear motion instant

JNeiger commented on soccer/Robot.cpp:
```diff
@@ -178,8 +170,11 @@ void OurRobot::moveTuning(Geometry2d::Point goal, float endSpeed) {
+    Geometry2d::Point targetVel = (goal - pos()).normalized() * endSpeed;
+    RobotInstant goal_instant{Geometry2d::Pose{targetPoint, 0},
+                              Geometry2d::Twist{targetVel, 0}, RJ::Time{0s}};
+    setMotionCommand(Planning::PathTargetCommand{goal_instant});
```
This is going from TuningPathCommand -> PathTargetCommand. I haven't looked, but I imagine they're different

kylestach replied:
We're removing the old `TuningPathCommand` for now. I'll add a TODO for it.

JNeiger commented on soccer/Robot.cpp:
```diff
@@ -196,28 +191,30 @@ void OurRobot::move(Geometry2d::Point goal, Geometry2d::Point endVelocity) {
-    intent().motion_command = std::make_unique<Planning::PathTargetCommand>(
-        MotionInstant(goal, endVelocity));
+    RobotInstant goal_instant;
+    goal_instant.pose = Pose{goal, 0};
```
This sets the heading target to 0 for every normal move?

kylestach replied:
Heading for this was ignored. I'm changing it to a linear motion instant to represent that.

JNeiger commented on soccer/Robot.cpp:
```diff
@@ -196,28 +191,30 @@ void OurRobot::move(Geometry2d::Point goal, Geometry2d::Point endVelocity) {
 }
 
-void OurRobot::settle(const std::optional<Point>& target) {
+void OurRobot::settle() {
```
Settle one touch thing

JNeiger commented on soccer/planning/PlannerNode.cpp:
```diff
@@ -1,99 +1,132 @@
-        context_->debug_drawer.drawShape(shape, Qt::black, "Global Obstacles");
+        RobotInstant start{robot.pose, robot.velocity, robot.timestamp};
+
+        // TODO(Kyle): Put motion constraints in intent.
```
todo issue #

JNeiger commented on soccer/planning/PlannerNode.hpp:
```diff
@@ -1,40 +1,36 @@
+
+private:
+    std::vector<std::unique_ptr<Planner>> planners_;
+    int planner_idx_;
```
What is this used for?

kylestach replied:
Left in accidentally. Just deleted it.

JNeiger commented on soccer/planning/PlannerNode.cpp:
```diff
@@ -1,99 +1,132 @@
+    planners_.push_back(std::make_unique<SettlePlanner>());
+    planners_.push_back(std::make_unique<CollectPlanner>());
+    planners_.push_back(std::make_unique<LineKickPlanner>());
+    planners_.push_back(std::make_unique<PivotPathPlanner>());
```
Does this order matter in this section? Or is it just the empty planner must be last?

kylestach replied:
Yeah. It only matters when multiple planners could be used to plan for the same command type. Right now, the only place that happens is with the empty/escape obstacles planner.

kylestach committed:
Merge branch 'mp-rewrite/trajectory' into mp-rewrite/plan-request

kylestach committed:
Address review comments

JNeiger reviewed (approved):
LGTM

kylestach committed:
Merge branch 'mp-rewrite/trajectory' into mp-rewrite/plan-request

oswinso reviewed (approved):
lgtm

kylestach committed:
[MP Rewrite Part 4] Motion Planning Primitives (#1496)

[MP Rewrite Part 4] Motion Planning Primitives

Primitives for motion planning including angle planning, RRT, velocity
profiling, Bezier fitting, and replanning. This is anticipated to be the
largest portion of the motion planning rewrite.

This is not expected to pass CI.

* Remove disabled for-loop in Processor

* Fix TODOs and docs in planning primitives

* Address review comments

* Make off-path threshold configurable

[MP Rewrite Part 5] Port previous planners to new framework (#1497)

Most of these are not functional changes from the old planners, but
rather simple ports to the new primitives. The biggest changes are that
we now use `Replanner` static functions and manually manage our own
"previous trajectory" objects instead of having "behavior-level"
planners (i.e. settle) own instances of an RRT planner and interacting
with it through plan requests.

This is the final PR in the rewrite, and it should pass CI. At the point
where it passes CI and all PRs are ready for merge, each PR should be
merged into its parent in sequence.

* Address review comments

* Move applyLowPassFilter into common, add TODOs

* Enable escape-obstacles planner

[MP Rewrite Part 6] Cleanup and fixes (#1512)

* Cleanup and fixes to make code run properly

* Fix test compilation and disable a trapezoidal motion profile edge case.

Some planner tests are still failing. This will be addressed in a later
commit.

* Prettify

* Make e2e planner tests pass.

Because they generate random scenarios, and the RRT planner is not
guaranteed to converge in a finite number of iterations (and the Bezier
approximation may be incorrect), these tests are currently flaky. That
should be fixed before this is merged to staging.

* Tidy fixes

* Linting

* Formatting

* Remove deleted file from CMakeLists, remove reuse condition from settle

* Don't replan in Collect::control

* Optimizations to geometry methods

* Reset and regenerate sim config

This deletes unnecessary variables and moves planning-related variables
into PathPlanner

* Formatting

* Add TODO in EscapeObstacles

* Tidy fixes

* Revert inconsistent clang-tidy recommended fix

* Temporarily disable flaky planner test

* Add NOLINTNEXTLINE

* Remove old sim config and add TODOs for trajectory utils
